appcelerator/titanium_mobile
# feat: add ability to run in background without UI

Work-in-Progress
---

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-27896

**Alloy:**
This PR requires the below Alloy PR to be merged into `appc` CLI 8.1.0.
https://github.com/appcelerator/alloy/pull/959

**Play Services:**
This PR requires the below "ti.playservices" PR.
https://github.com/appcelerator-modules/ti.playservices/pull/39

**Summary:**
- Added new "tiapp.xml" boolean property "run-in-background".
  * When set `true`, JS runtime is bound to the lifetime of the app process. Will be started as soon as app process starts, before UI.
  * When set `false` (the default), JS runtime is bound to the lifetime of the UI.
- Added new read-only `Ti.UI.hasSession` boolean property.
  * Will be `true` if app can host UI and create windows.
  * Will be `false` if app is unable to host any UI yet and is in the background.
- Added new `Ti.UI` event "sessionbegin".
  * Fired when app should create its windows.
  * Only fired when backgrounding feature is enabled.
- Bootstrap changes:
  * Added support for optional `showUI()` async method. Will be invoked on every "sessionbegin" event. (To be used by "ti.playservices" module.)
  * Added promises support to bootstrap `execute()` and `showUI()` functions.
- Added `Ti.Android.startForegroundServices()` method.
  * Needed to start a foreground service while app is in the background.
- Modified `Ti.App.Android.launchIntent` to never return `null`.
  * Now returns main launcher intent as a fallback.
  * Needed when creating an intent in the background without any UI.
- Add ability to start Android app in background after device bootup.
  * Feature automatically works if you add `RECEIVE_BOOT_COMPLETED` to "AndroidManifest.xml".
- Updated "ti.playservices" module:
  * Now uses Google Play Services library version `17.3.0`.
  * Now uses native update dialog when calling `makeGooglePlayServicesAvailable()`.
  * Updated bootstrap script to use this PR's async `showUI()` method.

---
**To-Do:**
- Test with "remote" policy.
- Add unit tests.
- Updates docs.

---
**Background Test:**
1. Create a new Classic App project.
2. Add the below "tiapp.xml" settings to the project.
3. Replace project's "app.js" file with the below.
4. Copy below "MyServices.js" file to project's "Resources" directory.
5. Build and run on Android.
6. Verify that the window shows "Launch Count: 1".
7. Verify that a "Foreground Service" notification appeared in status bar.
7. Back out of the app.
8. In the log, verify timestamps are posted once a second.
9. Re-launch the app.
10. Verify that the window shows "Launch Count: 2".
11. Reboot the Android device.
12. Wait for the "Foreground Service" notification to re-appear in status bar.
13. Tap on the notification.
14. Verify that it launches the app and it shows "Launch Count: 1".

tiapp.xml
```xml
<ti:app>
	<property name="run-in-background" type="bool">true</property>
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
			<uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
		</manifest>
		<services>
			<service url="MyService.js"/>
		</services>
	</android>
</ti:app>
```

app.js
```javascript
var count = 0;
function openWindow() {
	count++;
	var window = Ti.UI.createWindow();
	window.add(Ti.UI.createLabel({ text: "Launch Count: " + count }));
	window.open();
}

Ti.UI.addEventListener("sessionbegin", function() {
	openWindow();
});
if (Ti.UI.hasSession) {
	openWindow();
}

Ti.Android.startForegroundService(Ti.Android.createServiceIntent({ url: "MyService.js" }));

setInterval(function() {
	var date = new Date();
	Ti.API.info(String.format("@@@ %02d:%02d:%02d", date.getHours(), date.getMinutes(), date.getSeconds()));
}, 1000);
```

MyService.js
```javascript
Ti.API.info("@@@ Posting notification.");
Ti.Android.currentService.foregroundNotify(123, Ti.Android.createNotification({
	contentTitle: "Foreground Service",
	contentText: "Content Text",
	contentIntent: Ti.Android.createPendingIntent({
		intent: Ti.App.Android.launchIntent,
	}),
}));
```

---
**Test Restore After OS Force Quit:**
1. On Android, go to its "Developer Options" screen.
2. Tap on the "Background process limit" row. (It's near the bottom of the list.)
3. Tap on "No background processes". _(This tells the OS to force quit the last displayed app when launching another app.)_
4. Build and run the below code on the same Android device.
5. In the app, tap on the "Show Child Window" button.
6. Press the Android "Home" button.
7. Tap on another app such as "Gmail".
8. Press the Android "Home" button.
9. Resume the Titanium app.
10. Verify that the app restarts, starting with the splash screen. _(It will launch with a blank window, but quickly restart. Unfortunately, we are stuck with this behavior.)_
11. Change "tiapp.xml" property "run-in-background" to `false`.
12. Repeat steps 4-10.
13. Change device's "Background process limit" setting back to "Standard".

tiapp.xml
```xml
<ti:app>
	<property name="run-in-background" type="bool">true</property>
</ti:app>
```

app.js
```javascript
function openWindow() {
	var window = Ti.UI.createWindow();
	var openButton = Ti.UI.createButton({ title: "Show Child Window" });
	openButton.addEventListener("click", function(e) {
		var childWindow = Ti.UI.createWindow({ backgroundColor: "orange" });
		var closeButton = Ti.UI.createButton({ title: "Close Me" });
		closeButton.addEventListener("click", function(e) {
			childWindow.close();
		});
		childWindow.add(closeButton);
		childWindow.open();
	});
	window.add(openButton);
	window.open();
}

Ti.UI.addEventListener("sessionbegin", function() {
	openWindow();
});
if (Ti.UI.hasSession) {
	openWindow();
}
```
